internal/evaluator/rego.go

```diff
@@ -0,0 +1,134 @@
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"fmt"
+	"io"
+
+	"github.com/google/go-containerregistry/pkg/authn"
+	"github.com/google/go-containerregistry/pkg/name"
+	gcr "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/rego"
+	"github.com/open-policy-agent/opa/types"
+	log "github.com/sirupsen/logrus"
+)
+
+type key string
+
+const remoteLayerKey key = "ec.evaluator.remote.layer"
+
+type remoteLayerFn func(ref name.Digest, options ...remote.Option) (gcr.Layer, error)
+
+func remoteLayer(ctx context.Context, ref name.Digest, options ...remote.Option) (gcr.Layer, error) {
+	fn, ok := ctx.Value(remoteLayerKey).(remoteLayerFn)
+	if ok && fn != nil {
+		return fn(ref, options...)
+	}
+	return remote.Layer(ref, options...)
+}
+
+func WithRemoteLayerFn(ctx context.Context, fn remoteLayerFn) context.Context {
+	return context.WithValue(ctx, remoteLayerKey, fn)
+}
+
+func registerFetchOCIBlob() {
+	decl := rego.Function{
+		Name: "ec.fetch_oci_blob",
+		Decl: types.NewFunction(
+			types.Args(
+				types.Named("ref", types.S).Description("OCI blob reference"),
+			),
+			types.Named("blob", types.S).Description("the OCI blob"),
+		),
+		Memoize:          true,
+		Nondeterministic: true,
+	}
+
+	rego.RegisterBuiltin1(&decl, fetchOCIBlob)
+}
+
+const maxBytes = 10 * 1024 * 1024 // 10 MB
+
+func fetchOCIBlob(bctx rego.BuiltinContext, a *ast.Term) (*ast.Term, error) {
+	uri, ok := a.Value.(ast.String)
+	if !ok {
+		return nil, nil
+	}
+
+	ref, err := name.NewDigest(string(uri))
+	if err != nil {
+		log.Errorf("fetchOCIBlob new digest: %s", err)
+		return nil, nil
+	}
+
+	opts := []remote.Option{
+		remote.WithTransport(remote.DefaultTransport),
+		remote.WithContext(bctx.Context),
+		remote.WithAuthFromKeychain(authn.DefaultKeychain),
+	}
+
+	rawLayer, err := remoteLayer(bctx.Context, ref, opts...)
+	if err != nil {
+		log.Errorf("fetchOCIBlob remote layer: %s", err)
+		return nil, nil
+	}
+
+	layer, err := rawLayer.Uncompressed()
+	if err != nil {
+		log.Errorf("fetchOCIBlob layer uncompressed: %s", err)
+		return nil, nil
+	}
+	defer layer.Close()
+
+	// Other algorithms are technically supported, e.g. sha512. However, support for those is not
+	// complete in the go-containerregistry library, e.g. name.NewDigest throws an error if sha256
+	// is not used. This is good for now, but may need revisiting later.
+	hasher := sha256.New()
+	// Setup some safeguards. First, use LimitReader to avoid an unbounded amount of data from being
+	// read. Second, use TeeReader so we can compute the digest of the content read.
+	reader := io.TeeReader(io.LimitReader(layer, maxBytes), hasher)
+
+	var blob bytes.Buffer
+	if _, err := io.Copy(&blob, reader); err != nil {
+		log.Errorf("fetchOCIBlob copy buffer: %s", err)
+		return nil, nil
+	}
+
+	sum := fmt.Sprintf("sha256:%x", hasher.Sum(nil))
+	if sum != ref.DigestStr() {
```

Comment:
Is this because LimitReader might truncate the stream? If not so, think the [fetcher](https://github.com/google/go-containerregistry/blob/a54d64203cffcbf94146e04069aae4a97f228ee2/pkg/v1/remote/fetcher.go#L276) already checks the digest.

Reply:
That's exactly why. I wish LimitReader would return whether or not content was truncated. If it did, this wouldn't be needed.

Reply:
Added a comment to make it more obvious.